Content.Shared/Weapons/Reflect/ReflectSystem.cs

```diff
@@ -49,72 +49,128 @@ public override void Initialize()
+            return;
 
-            args.Cancelled = true;
-            break;
-        }
+        if (!UserCanReflect(userUid, out var bestReflectorUid))
+            return;
+
+        if (!TryReflectProjectile(userUid, bestReflectorUid.Value, args.ProjUid))
+            return;
+
+        args.Cancelled = true;
     }
 
-    private void OnReflectCollide(EntityUid uid, ReflectComponent component, ref ProjectileReflectAttemptEvent args)
+    private void OnObjectReflectHitscanAttempt(EntityUid objectUid, ReflectComponent component, ref HitScanReflectAttemptEvent args)
+    {
+        if (args.Reflected || (component.Reflects & args.Reflective) == 0x0)
+            return;
+
+        if (!TryReflectHitscan(objectUid, objectUid, args.Shooter, args.SourceItem, args.Direction, out var dir))
+            return;
+
+        args.Direction = dir.Value;
+        args.Reflected = true;
+    }
+
+    private void OnObjectReflectProjectileAttempt(EntityUid objectUid, ReflectComponent component, ref ProjectileReflectAttemptEvent args)
     {
         if (args.Cancelled)
             return;
 
-        if (TryReflectProjectile(uid, uid, args.ProjUid, reflect: component))
-            args.Cancelled = true;
+        if (!TryReflectProjectile(objectUid, objectUid, args.ProjUid))
+            return;
+
+        args.Cancelled = true;
     }
 
-    private bool TryReflectProjectile(EntityUid user, EntityUid reflector, EntityUid projectile, ProjectileComponent? projectileComp = null, ReflectComponent? reflect = null)
+    /// <summary>
+    /// Can a user reflect something that's hit them? Returns true if so, and the best reflector available in the user's equipment.
+    /// </summary>
+    private bool UserCanReflect(EntityUid uid, [NotNullWhen(true)] out EntityUid? bestReflectorUid, EntityUid? projectile = null)
+    {
+        bestReflectorUid = null;
+        ReflectComponent? bestReflector = null;
+        ReflectiveComponent? reflective = null;
+
+        // If it's a projectile trying to hit someone, make sure it can be reflected.
+        if (projectile != null && !TryComp(projectile, out reflective))
+            return false;
+
+        foreach (var entityUid in _inventorySystem.GetHandOrInventoryEntities(uid, SlotFlags.WITHOUT_POCKET))
+        {
+            if (!TryComp<ReflectComponent>(entityUid, out var comp))
+                continue;
+
+            if (!comp.Enabled)
+                continue;
+
+            if (bestReflector != null && bestReflector.ReflectProb >= comp.ReflectProb)
+                continue;
+
+            if (reflective != null && (comp.Reflects & reflective.Reflective) == 0x0)
+                continue;
+
+            bestReflector = comp;
+            bestReflectorUid = entityUid;
+        }
+
+        return bestReflectorUid != null && bestReflector != null;
+    }
+
+    private bool TryReflectProjectile(EntityUid user, EntityUid reflector, EntityUid projectile, ProjectileComponent? projectileComp = null)
```

Comment:
Make the methods in this system use Entity<> to indicate what component corresponds to what uid

Reply:
Yeah just grokked how Entity<> works on the waddle PR, will implement those learnings here